course/models.py

```diff
@@ -1704,16 +1721,60 @@ def consume(self, iterable, set_is_superseded=False):
             gchange.is_superseded = False
             self._consume_grade_change(gchange, set_is_superseded)
 
-        valid_grade_changes = sorted(
-                (gchange
-                for gchange in self.attempt_id_to_gchange.values()
-                if gchange.percentage() is not None),
-                key=lambda gchange: gchange.grade_time)
+        gchanges = [gchange for gchange in self.attempt_id_to_gchange.values()
+                   if gchange.percentage() is not None]
+
+        valid_grade_changes = sorted(gchanges, key=lambda x: (x.grade_time, x.pk))
+
+        import copy
+        _valid_percentages_copy = copy.copy(self.valid_percentages)
 
         self.valid_percentages.extend(
                 cast(GradeChange, gchange.percentage())
                 for gchange in valid_grade_changes)
 
+        # {{{ Calculate the earliest percentage and latest percentage, instead of
+        # get them from self.valid_percentages, so as to avoid inconsistent results.
+        # See https://github.com/inducer/relate/pull/423#discussion_r162121467
```

Comment:
That link does not work.

Reply:
Oh, that link is targeting our previous discussion on the potential inconsisitency might araise when sorting `GradeChange` objects with/without a session.

Reply:
(The discussion is collapsed, and that link can't uncollpase it).